test_python/test_tc.py

```diff
@@ -12,89 +12,499 @@
+        I = Variable(torch.randn(N, G, D, H, W).cuda())
+        gamma = Variable(torch.randn(G, D).cuda())
+        beta = Variable(torch.randn(G, D).cuda())
+
+        print("\nTune from naive options, don't store best options")
+        options = group_norm.tune(
+            I,
+            gamma,
+            beta,
+            starting_options=tc.MappingOptions('naive'),
+            tuner_config=tuner_config)
+        print("\nRun best tuned options")
+        Sum, SumSq, O = group_norm(I, gamma, beta)
+        # TODO: test results!!!
+
+    #
+    # This test implements group normalization as 2 TC kernels using the
+    # tc.define abstraction. In particular it is possible to insert reshapes
+    # between TCs and have moments operate on a 2-D view which is mapped
+    # more efficiently to CUDA at this time
+    # (Note: loop collapsing modulo conforming strides is an easy TC-level
+    #  transformation with nice expected benefits).
+    #
+    def test_group_norm_2kernels(self):
+        group_normalization = """
+            def moments(float(N, K) I) -> (mean, var) {
+                # var = E(x^2) - mean^2.
+                mean(n) +=! I(n, r_k)
+                 var(n) +=! I(n, r_k) * I(n, r_k)
+                mean(n)  = mean(n) / (K)
+                 var(n)  =  var(n) / (K) - mean(n) * mean(n)
+            }
+
+            def group_normalization(
+                float(N, G, D, H, W) I, float(G, D) gamma, float(G, D) beta,
+                float(N, G) mean, float(N, G) var) -> (O)
+            {
+                O(n, g, d, h, w) = gamma(g, d)
+                    * ( I(n, g, d, h, w) - mean(n, g) )
+                    * rsqrt( var(n, g) + 1e-5 )
+                    + beta(g, d)
+            }
+        """
+
+        N, G, D, H, W = 32, 32, 4, 56, 56
+        moments = tc.define(
+            group_normalization,
+            entry_point="moments")
+        group_norm = tc.define(
+            group_normalization,
+            entry_point="group_normalization")
+        I = Variable(torch.randn(N, G, D, H, W).cuda())
+        gamma = Variable(torch.randn(G, D).cuda())
+        beta = Variable(torch.randn(G, D).cuda())
+
+        print("\nTune moments from naive options, don't store best options")
+        moments.tune(
+            I.view((N * G, -1)),
+            starting_options=tc.MappingOptions('naive'),
+            tuner_config=tuner_config)
+        print("\nRun tuned moments with best options")
+        mean, var = moments(I.view((N * G, -1)))
 
-PATH_PREFIX = os.path.join(os.path.dirname(os.path.abspath(__file__)), "tc_test")
+        print("\nTune group_norm from naive options, don't store best options")
+        group_norm.tune(
+            I,
+            gamma,
+            beta,
+            mean.view((N, G)),
+            var.view((N, G)),
+            starting_options=tc.MappingOptions('naive'),
+            tuner_config=tuner_config)
+        print("\nRun tuned group_norm with best options")
```

Comment:
I'd generally advise against using `print`s in tests. Their whole point is to be automatic, so they won't help you too much.

Reply:
mostly inherited from older impl, killed, thanks!